File: Arrays/Second_Largest/optimal_sol.cpp
#include<bits/stdc++.h>
using namespace std;

int secondLargest(vector<int> a, int n){
    if(n < 2)
        return -1;
    
    int large = INT_MIN, second_large = INT_MIN;
    for(int i=0; i<n; i++){
        if(a[i] > large){
            second_large = large;
            large = a[i];
        }
        else if (a[i] > second_large && a[i] != large){
            second_large = a[i];
        }
    }
    return second_large;
}

int secondSmallest(vector<int> a, int n){
    if (n < 2)
        return -1;
    int small = INT_MAX, second_small = INT_MAX;
    for(int i=0; i<n; i++){
        if(a[i] < small){
            second_small = small;
            small = a[i];
        }
        else if(a[i] < second_small && a[i] != small){
            second_small = a[i];
        }

    }
    return second_small;

}

vector<int> getSecondOrderElements(int n, vector<int> a){
    int slargest = secondLargest(a,n);
    int ssmallest = secondSmallest(a,n);
    return {slargest, ssmallest};
}

int main(){
    vector<int> a = {1, 2, 4, 6, 7, 5};
    int n = a.size();
    vector<int> result = getSecondOrderElements(n,a);
    cout << "Second largest element is: " << result[0] << endl;
    cout << "Second smallest element is: " << result[1] << endl;
    return 0;
}

File: Arrays/Second_Largest/optimal_sol.py
def second_Largest(arr):
    if len(arr) < 2:
        return None
    largest = float('-inf')
    slargest = float('-inf')
    for i in range(len(arr)):
        if arr[i] > largest:
            slargest = largest
            largest = arr[i]
        elif arr[i] > slargest and arr[i] != largest:
            slargest = arr[i]

    return slargest

def second_Smallest(arr):
    if len(arr) < 2:
        return None
    smallest = float('inf')
    ssmallest = float('inf')
    for i in range(len(arr)):
        if arr[i] < smallest:
            ssmallest = smallest
            smallest = arr[i]
        elif arr[i] < ssmallest and arr[i] != smallest:
            ssmallest = arr[i]
    return ssmallest

def getSecondOrderElement(arr):
    second_smallest = second_Smallest(arr)
    second_largest = second_Largest(arr)

    print("Second Smallest element: ", second_smallest)
    print("Second Largest element: ", second_largest)



if __name__ == "__main__":
    arr = [2,4,6,8,9]

    getSecondOrderElement(arr)


            